2FANotes.txt

```diff
@@ -0,0 +1,55 @@
+
+This error is coming from
+
+src\BlazorIdentity.Server\BlazorServerSignInManager.cs line 67
+
+var result = await base.PasswordSignInAsync(userName, password, isPersistent, lockoutOnFailure);
+
+Exception
+
+blazor.server.js:1 
+        
+       [2022-07-13T13:32:27.080Z] Error: System.InvalidOperationException: OnStarting cannot be set because the response has already started.
+   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ThrowResponseAlreadyStartedException(String value)
+   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.OnStarting(Func`2 callback, Object state)
+   at Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler.InitializeHandlerAsync()
+   at Microsoft.AspNetCore.Authentication.AuthenticationHandler`1.InitializeAsync(AuthenticationScheme scheme, HttpContext context)
+   at Microsoft.AspNetCore.Authentication.AuthenticationHandlerProvider.GetHandlerAsync(HttpContext context, String authenticationScheme)
+   at Microsoft.AspNetCore.Authentication.AuthenticationService.AuthenticateAsync(HttpContext context, String scheme)
+   at Microsoft.AspNetCore.Identity.SignInManager`1.IsTwoFactorClientRememberedAsync(TUser user)
+   at Microsoft.AspNetCore.Identity.SignInManager`1.CheckPasswordSignInAsync(TUser user, String password, Boolean lockoutOnFailure)
+   at Microsoft.AspNetCore.Identity.SignInManager`1.PasswordSignInAsync(TUser user, String password, Boolean isPersistent, Boolean lockoutOnFailure)
+   at Microsoft.AspNetCore.Identity.SignInManager`1.PasswordSignInAsync(String userName, String password, Boolean isPersistent, Boolean lockoutOnFailure)
+   at BlazorIdentity.Server.BlazorServerSignInManager`1.BlazorIdentity.IBlazorSignInManager<TUser>.PasswordSignInAsync(String userName, String password, Boolean isPersistent, Boolean lockoutOnFailure) in D:\projects\BlazorIdentityFork\src\BlazorIdentity.Server\BlazorServerSignInManager.cs:line 67
+   at BlazorIdentity.UI.Login`1.HandleSubmit() in D:\projects\BlazorIdentityFork\src\BlazorIdentity.UI\Login.razor:line 58
+   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
+   at Microsoft.AspNetCore.Components.Forms.EditForm.HandleSubmitAsync()
+   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
+   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle, ComponentState owningComponentState)
+log @ blazor.server.js:1
+
+I read an interesting note here:
+
+https://code-maze.com/two-step-verification-aspnet-core-identity/
+
+It is important to mention that as soon as a user gets redirected to the LoginTwoStep action, the new Identity.TwoFactorUserId cookie will be created in our browser. This cookie contains important data about the current user.
+
+I am suspicious this has something to do with what is going on. 
```

Comment:
Yeah you can't write out cookies from the Blazor Hub connection so the Blazor Server specific bits in here have to be updated to handle the 2fa login explicitly in a similar way to the normal sign in. I'll take a look.